lib/plug/crypto/message_encryptor.ex

```diff
@@ -19,76 +19,264 @@ defmodule Plug.Crypto.MessageEncryptor do
+  end
+
+  # Verifies and decrypts a message using AES128-CBC mode
+  # with HMAC-SHA-1 for the authentication code.
+  # 
+  # Decryption will never be performed prior to verification.
+  defp aes256_cbc_hmac_sha1_decrypt(cipher_text, secret, sign_secret)
+      when bit_size(secret) > 256,
+    do: aes256_cbc_hmac_sha1_decrypt(cipher_text, binary_part(secret, 0, 32), sign_secret)
+  defp aes256_cbc_hmac_sha1_decrypt(cipher_text, secret, sign_secret)
+      when is_binary(cipher_text)
+      and bit_size(secret) === 256
+      and is_binary(sign_secret) do
+    case decode_legacy_token(cipher_text, sign_secret) do
+      {"A256CBC-HS1", _encrypted_key, iv, cipher_text, cipher_tag}
+          when bit_size(iv) === 128 and bit_size(cipher_tag) === 160 ->
+        key = secret
+        :crypto.block_decrypt(:aes_cbc256, key, iv, cipher_text)
+        |> case do
+          plain_text when is_binary(plain_text) ->
+            pkcs7_unpad(plain_text)
+          _ ->
+            :error
+        end
+      _ ->
+        :error
+    end
+  end
+
+  # Encrypts and authenticates a message using AES128-GCM mode.
+  # 
+  # A random 128-bit content encryption key (CEK) is generated for
+  # every message which is then encrypted with `aes_gcm_key_wrap/3`.
+  defp aes128_gcm_encrypt(plain_text, secret, sign_secret)
+      when bit_size(secret) > 256,
+    do: aes128_gcm_encrypt(plain_text, binary_part(secret, 0, 32), sign_secret)
+  defp aes128_gcm_encrypt(plain_text, secret, sign_secret)
+      when is_binary(plain_text)
+      and bit_size(secret) in [128, 192, 256]
+      and is_binary(sign_secret) do
+    key = :crypto.strong_rand_bytes(16)
+    iv = :crypto.strong_rand_bytes(12)
+    aad = "A128GCM"
+    {cipher_text, cipher_tag} = :crypto.block_encrypt(:aes_gcm, key, iv, {aad, plain_text})
+    encrypted_key = aes_gcm_key_wrap(key, secret, sign_secret)
+    encode_token(aad, encrypted_key, iv, cipher_text, cipher_tag)
+  end
+
+  # Verifies and decrypts a message using AES128-GCM mode.
+  # 
+  # Decryption will never be performed prior to verification.
+  # 
+  # The encrypted content encryption key (CEK) is decrypted
+  # with `aes_gcm_key_unwrap/3`.
+  defp aes128_gcm_decrypt(cipher_text, secret, sign_secret)
+      when bit_size(secret) > 256,
+    do: aes128_gcm_decrypt(cipher_text, binary_part(secret, 0, 32), sign_secret)
+  defp aes128_gcm_decrypt(cipher_text, secret, sign_secret)
+      when is_binary(cipher_text)
+      and bit_size(secret) in [128, 192, 256]
+      and is_binary(sign_secret) do
+    case decode_token(cipher_text) do
+      {aad = "A128GCM", encrypted_key, iv, cipher_text, cipher_tag}
+          when bit_size(iv) === 96 and bit_size(cipher_tag) === 128 ->
+        encrypted_key
+        |> aes_gcm_key_unwrap(secret, sign_secret)
+        |> case do
+          {:ok, key} ->
+            :crypto.block_decrypt(:aes_gcm, key, iv, {aad, cipher_text, cipher_tag})
+          _ ->
+            :error
+        end
+        |> case do
```

Comment:
The current approach is fine. :) I just got confused for a while because I though block_decrypt could never fail. We will be able to rewrite all of those slightly better once we use Elixir v1.3.